### ISSUE TITLE
OSError: [Errno 22] Invalid argument while running evaluation file

### ISSUE THREAD

**Issue by aakdani** [mvrl/sink-seg#1]
Hello, I have been trying to implement your code as we want to work in a similar direction, however, I came across the following error while running the Evaluation file;

" File "C:\Users\danish\anaconda3\envs\sink-holes\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)

OSError: [Errno 22] Invalid argument

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\danish\anaconda3\envs\sink-holes\lib\multiprocessing\spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "C:\Users\danish\anaconda3\envs\sink-holes\lib\multiprocessing\spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
_pickle.UnpicklingError: pickle data was truncated "

Would you please help me identify what is causing this issue?

Thank you!

**Comment by jacobsn** [comment]
It's hard to help when I don't know what code you are running. Also, did you make any changes to the code or any parameters?

**Comment by aakdani** [comment]
I'm running the files provided in your repository and I was able to run the train.py file successfully but the evaluate.py file is giving me this error. I haven't made any changes to any of the code or parameters used in the code so far. Does it have something to do with the memory in the data loader?

**Comment by jacobsn** [comment]
It would be helpful if I knew what line of code in evaluate.py was causing the problem.

**Comment by aakdani** [comment]
Its line 198 "for i, data in enumerate(data_loader_test) ", I checked and looping throught "data_loader_test"  is causing the error. I also tried changing the "cfg.train.num_workers" to  1 from 4 but still getting the error. So I'm not sure if it is due to the memory or if it is some other issue.

**Comment by jacobsn** [comment]
I'm not sure, I'll let @Usman-Rafique jump in since he knows the code better.

I'd suggest setting num_workers to zero to see if it's just the multiprocessing problem, or if the multiprocessing is hiding the real error.

**Comment by aakdani** [comment]
You are right, I think there is some problem in multiprocessing and multi-processing in masking the problem, I will set num_workers to zero and run the code. Hopefully, it solves the problem. Thank you for your quick response.

**Comment by Usman-Rafique** [comment]
Hi @aakdani, I just took a quick look and the code structure for `train` and `evaluate` seems the same. Also, the train and val/test dataloaders have pretty much the same settings. Please update here once you try with `num_workers = 0`.

Since you pointed out that the error is in line 198, that means the problem is only in the test dataloader, and the val loader loop (starting at [L101](https://github.com/mvrl/sink-seg/blob/master/evaluate.py#L101)) was successfully executed. 

Did you see any error, warning, or any odd numbers being printed when the dataloaders are being prepared (and split to train/val/test)? Once again, importantly, did you change any settings, such as `split_ratio` at (L293)[https://github.com/mvrl/sink-seg/blob/master/data_factory.py#L293]?

**Comment by aakdani** [comment]
I will update here once I try with num_workers = 0. I got the same error with the val dataloader loop but I tried running the code couple of times and it finally executed but the test dataloader never executed. 
No I haven't came across any error  when the dataloaders are being prepared (and split to train/val/test) but when the dataloader loop is executing, sometimes I get the warning related to setting the datatype int64 but I don't think that could be the issue. 
In regard to settings, I didn't change any settings in code so far. They are exactly the same as in github repository. I will post an update after running the script with num_worker = 0, I strongly believe that there is some memory issue with multiprocessing so that could be the reason.
Thank you for your response. It was really helpful.
